option without value can prompt or use default

amy-lei opened the pull request:
Resolves #549 
Resolves #736
Resolves #764 
Resolves #921 
Resolves #1015 

Add `prompt_required` attribute to `Option` so that when `prompt_required=False`, the user will only be prompted for an input if the option is specified without a value. The attribute defaults to `True` to maintain the current behavior of option prompts.

`is_flag=False` now means that providing the value to the option is optional. If the flag is provided without the value, it can be prompted for or use `flag_value`.

amy-lei commented on src/click/parser.py:
```diff
@@ -337,11 +337,27 @@ def _match_long_opt(self, opt, explicit_value, state):
+            # user will be prompted for an input if no argument is passed.
             nargs = option.nargs
             if len(state.rargs) < nargs:
-                _error_opt_args(nargs, opt)
```
I've noticed this block of code in `_match_long_opt` is identical to the one in `_match_short_opt`. So when making changes, I kept it this way and added the changes twice. Should this be factored out or is it not too big of a deal?

amy-lei commented on src/click/parser.py:
```diff
@@ -337,11 +337,27 @@ def _match_long_opt(self, opt, explicit_value, state):
+                    _error_opt_args(nargs, opt)
             elif nargs == 1:
-                value = state.rargs.pop(0)
+                next_rarg = state.rargs[0]
```
I've noticed in the current codebase that this code block in `_match_long_opts` is identical in `_match_short_opts`. I kept it this way when making changes and just added the changes in both places. Should this be factored out, or is it not that significant? 

https://github.com/pallets/click/blob/637ea206c348100582beda3c76a8c51e9d7437ab/src/click/parser.py#L341-L347

davidism commented:
I'm wondering if it's appropriate to have `prompt` get triggered from the parser, it seems like the parser should only be concerned with the command line, and prompting should happen during later processing.

amy-lei commented:
> I'm wondering if it's appropriate to have `prompt` get triggered from the parser, it seems like the parser should only be concerned with the command line, and prompting should happen during later processing.

I moved it into `full_process_value` since that's where it originally prompts for a value. But to do so, I had to add an optional param to let it know there was a flag present

davidism commented:
I think there's another issue about options where the value is optional, maybe this solution will help with that too.

davidism commented:
I was thinking of #549.

amy-lei commented:
> I was thinking of #549.

Just found it and was about to comment it too haha. Should I try to fit that into this PR too?

davidism commented:
It seems like it should be a natural fit with this, since we now track if only the flag was passed. If it's only a flag, and there is a default, use the default instead of prompting. I don't *think* that should collide with any existing combination of settings.

davidism commented on src/click/core.py:
```diff
@@ -1661,7 +1661,7 @@ def handle_parse_result(self, ctx, opts, args):
             value = self.consume_value(ctx, opts)
             try:
-                value = self.full_process_value(ctx, value)
+                value = self.full_process_value(ctx, value, is_flag=self.name in opts)
```
Options already have an `is_flag` attribute. Can we take advantage of that? This probably also ties in with #549.

amy-lei replied:
An idea I have is to set `is_flag=True` for prompt options that have `prompt_required=False`

amy-lei commented:
@davidism I've added the ability to use options with optional args via the `_flag_needs_value` attribute we talked about. I currently have it so that setting `prompt_required=False` implies `_flag_needs_value=True`, and this simplifies the checks in the parser. 

However, this did not get rid of the additional parameter we added to `full_process_value` because the only way to distinguish  between the value being `None` due to the parser setting it that way and it being `None` due to the user not passing it in all can't be made without the `opts` dict.

An idea in mind is to instead of having `prompt_required=False` imply `_flag_needs_value=True` in the constructor, maybe we could set it to True in the parser? And then set it back after we've prompted? This way we wouldn't need the additional parameter anymore. But I'm also not too fond about this since we're changing attribute of the Option instance midway.

davidism commented:
Instead of the parser producing `None`, it could produce a token that processing later on can check for.

```python
# define the sentinel value at the top of the module
flag_needs_value = object()

# return it when parsing
def match(...):
    return _flag_needs_value

# check it when processing
def process(...):
    if value is _flag_needs_value:
        ...
    elif value is None:
        ...
```

davidism commented:
I updated the PR description with the issues related to using a value instead of prompting. Also closed some existing PRs that will be handled by this. Let's make sure any tests they had are reflected here if applicable.

amy-lei commented:
Sounds good :+1: I'll be sure to add those tests and update the docs.

Some clarifying questions:
1) If the user sets `is_flag=False` but does not pass in a `flag_value`, should an error be raised during initialization? during parsing? Or should it be fine to just let the value default to something like `None` instead?
2) From the description, does this mean that in order to allow the user to prompt when no value is passed, `is_flag=False` must be explicitly passed too? Is is just passing `prompt_required=False` sufficient?

davidism commented:
1. For now it's better to require that `flag_value` be set. While we could always remove that restriction later, it would be harder to add it in.
2. For this, I think if `is_flag` is left as `None`, `prompt_required=False` can set it to `False`.

amy-lei commented:
I've used the sentinel value idea, although was not too sure about the name or if it was appropriate to name it the same name as the attribute. 

I also noticed #1230 enforced that the option cannot be `multiple` or have `nargs!=1`. I'm wondering if that should be the case here too or if it should be allowed instead. I did not add tests for it nor explicitly mention it in the docs yet.

davidism commented:
Will get to this after #1649. They're all related, might affect my answer to the question about `multiple` and `nargs`.

brendanator commented:
You may have already addressed this but I was attempting to implement something similar to the color option from git diff:
```man
       --color[=<when>]
           Show colored diff.  --color (i.e. without =<when>) is the same as
           --color=always.  <when> can be one of always, never, or auto. It can
           be changed by the color.ui and color.diff configuration settings.
```

The color option has a flag value of `always` and defaults to `never` so it will be covered by this functionality. 

The main difference I see is that the value must be separated from the option by `=`.  I think this is probably required to avoid the option gobbling up a parameter. e.g. `git diff --color path/to/file` should not set the color option to `path/to/file`

The example in the docs don't seem to handle this.

davidism commented:
Working on rebasing this after #1649.
